### PR TITLE
Add dialogue based creation of work items (#352)

### PR DESCRIPTION
**Pull request by nimishamukherjee**
Add dialogue based creation of work items (#352)

User story - https://github.com/almighty/almighty-core/issues/501
UX - https://redhat.invisionapp.com/share/Y39KFTAE7

### CODE REVIEW

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 1]
```diff
@@ -209,6 +212,8 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
       </div>
     </fieldset>
   </form>
-  <alm-work-item-link [workItem]="workItem" [loggedIn]="loggedIn"></alm-work-item-link>
-  <alm-work-item-comment [workItem]="workItem" [loggedIn]="loggedIn"></alm-work-item-comment>
-</div>
+  <div *ngIf="!addNewWI">
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Does this mean we can't add a link when we create a new workitem?

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 2]
```diff
@@ -17,7 +17,7 @@
             </li>
           </ul>
         </div>
-        <h5 class="detail-id pull-left" id="wi-detail-id">{{workItem.id}}</h5>
+        <h5 *ngIf="!addNewWI" class="detail-id pull-left" id="wi-detail-id">{{workItem.id}}</h5>
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Can we just do it like if workItem.id available or not? So maybe replace {{workItem.id}} with
{{workItem?.id}} and remove *ngIf="!addNewWI".

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html (hunk 1 above)
Comment (nimishamukherjee) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: As per the wireframe, all the fields remain disabled until the title has been entered and saved. On the UI, the comments and link are not shown until the title is created.
Implementation wise, 
1. Create a work item with type + title
2. On save, we get the work item ID
3. We reload the details page with the newly created work item (this is the same as edit now)

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 3]
```diff
@@ -66,7 +66,7 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
 
         <div class="col-md-4 col-sm-5 col-xs-5 wi-state-btn pull-right padding0">
           <div dropdown class="dropdown card-pf-time-frame-filter detail-status-dropdown">
-            <button dropdown-open 
+            <button dropdown-open [disabled]='addNewWI'
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Why do you need to keep it disabled until the item is added? I understand that changing it makes an API call for update but you can use the same logic in the event listener. In fact you can have all the fields like Description, States, Assignee, just don't let any of the vent listener call the save method until you have workItem.attributes['system.title']

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 4]
```diff
@@ -89,19 +89,19 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
       </div>
       <div class="col-md-12 col-sm-12 col-xs-12 detail-assignee-wrap">
           <span
-            *ngIf="!workItem.relationalData.assignees?.length"
+            *ngIf="addNewWI || !workItem.relationalData.assignees?.length"
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 5]
```diff
@@ -89,19 +89,19 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
...
             (click)="activeSearchAssignee()" 
             class="pull-left fa fa-user-plus user-assign-icon">
           </span>
           <img 
-            *ngIf="workItem.relationalData.assignees?.length"
+            *ngIf="!addNewWI && workItem.relationalData.assignees?.length"
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 6]
```diff
@@ -89,19 +89,19 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
...
             (click)="activeSearchAssignee()"
             class="pull-left user-assign-avatar"
             src="{{workItem.relationalData.assignees[0]?.attributes.imageURL + '&s=20'}}" 
             onError="this.src='https://avatars0.githubusercontent.com/u/563119?v=3&s=20'" />
           
           
-          <div class="pull-left detail-assignee-name">
+          <div *ngIf="!addNewWI" class="pull-left detail-assignee-name">
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 7]
```diff
@@ -148,14 +148,17 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
               </div>
           </div>
           <div class="pull-right">
-            <span title="{{workItem.attributes['system.created_at'] | date:'fullDate'}}">
-              Created at: {{workItem.attributes['system.created_at'] | almMomentTime}}
+            <span *ngIf="!addNewWI" title="{{workItem.attributes['system.created_at'] | date:'fullDate'}}">
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.
In this case you can just replace 'Oops do not have the time!' with 'Creating Now'

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 8]
```diff
@@ -148,14 +148,17 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
...
+            </span>
+            <span *ngIf="addNewWI" title="{{workItem.attributes['system.created_at'] | date:'fullDate'}}">
+              Creating now
             </span>
           </div>
       </div>
 
       <div class="col-md-12 col-sm-12 col-xs-12 detail-description-wrap">
-        <div class="col-md-12 col-sm-12 col-xs-12 detail-description" *ngIf="loggedIn">
+        <div class="col-md-12 col-sm-12 col-xs-12 detail-description" *ngIf="loggedIn && !addNewWI">
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 9]
```diff
@@ -191,7 +194,7 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
             </div>
           </div>
         </div>
-        <div id="wi-detail-desc" class="col-md-12 description" *ngIf="!loggedIn">
+        <div id="wi-detail-desc" class="col-md-12 description" *ngIf="!loggedIn || addNewWI">
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html (hunk 1 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Same logic as I explained up there. I don't think addNewWI flag is needed anywhere.

Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Well, this actually means we can't create a link or comment until the item is created. I guess this is fine.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 10]
```diff
@@ -174,7 +177,7 @@ <h2 *ngIf="loggedIn" id="wi-detail-title-click" class="detail-title truncate">{{
                 #desc 
                 (onUpdate)="descUpdate($event)"
                 (keydown.enter)="onUpdateDescription()"
-                (keyup.enter)="preventDef($event)"
+                (keyup.enter)="preventDef($event)"
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: Please remove the white space

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 11]
```diff
@@ -52,6 +52,8 @@ export class WorkItemDetailComponent implements OnInit {
 
   loggedInUser: any;
 
+  addNewWI: Boolean = false;
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: Probably you'll not need it.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 12]
```diff
@@ -72,24 +74,53 @@ export class WorkItemDetailComponent implements OnInit {
...
-            this.activeOnList(400);
-          })
-          .catch (err => this.closeDetails());
+        if (id.indexOf('new') >= 0){
+          //Add a new work item          
+          this.addNewWI = true;
+          let type = id.substring((id.indexOf('?') + 1), id.length);
+          this.createWorkItemObj(type);
+          this.getAllUsers();
+          this.activeOnList(400);
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: This line should not be here. This function requires a workItem ID which is not available at this point.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 13]
```diff
@@ -72,24 +74,53 @@ export class WorkItemDetailComponent implements OnInit {
...
-            this.getAllUsers();
-            this.activeOnList(400);
-          })
-          .catch (err => this.closeDetails());
+        if (id.indexOf('new') >= 0){
+          //Add a new work item          
+          this.addNewWI = true;
+          let type = id.substring((id.indexOf('?') + 1), id.length);
+          this.createWorkItemObj(type);
+          this.getAllUsers();
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: This function wouldn't be required following the code on template, but since I think choosing the assignee before saving the item should be implemented, this will be needed.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 14]
```diff
@@ -205,12 +236,20 @@ export class WorkItemDetailComponent implements OnInit {
   }
 
   save(): void {
-    this.workItemService
-      .update(this.workItem)
+    if (this.addNewWI){
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: This is only one place I think the flag is needed.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 15]
```diff
@@ -205,12 +236,20 @@ export class WorkItemDetailComponent implements OnInit {
   }
 
   save(): void {
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: I would recommend to separate save and create method for mainly two reason - 

- Better readability.
- In very near future we are going to support partial update, when this will become more complex.

actually let's discuss about it.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts (hunk 14 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: I just thought, even this can be avoided. 
Check for !this.workItem.id instead

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 16]
```diff
@@ -72,24 +74,53 @@ export class WorkItemDetailComponent implements OnInit {
...
+              this.activeOnList(400);
+            })
+            .catch (err => this.closeDetails());
+        }
       }
     });
   }
 
+  createWorkItemObj(type: string) {
+    this.workItem = new WorkItem();
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: Set this.workItem.id = null or false and use this as flag in places.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 17]
```diff
@@ -204,14 +235,23 @@ export class WorkItemDetailComponent implements OnInit {
...
+  save(): void {        
+    if (this.workItem.id){
+      this.workItemService
+        .update(this.workItem)
+        .then((workItem) => {
+          this.workItem.attributes['version'] = workItem.attributes['version'];
+          this.activeOnList();          
+      });
+    }else{
+      if (this.workItem.attributes['system.title']){
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: Can you use **this.validTitle** flag here. So that if you write some special validation logic tomorrow in **isValid()** function, then that'll work here as well.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 18]
```diff
@@ -1,6 +1,59 @@
 @import "../../../assets/stylesheets/color-variables";
 @import "../../../assets/stylesheets/custom";
 
+.add-detailed-wi{
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: This file needs indentation.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html (hunk 1 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: I'm ok with that, just wanted to confirm.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html (hunk 3 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: @sanbornsen that only works for Add but we want each to update in Edit mode.

Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: @joshuawilson yeah, so in case of adding, no API should be fired until the title is valid. So, she updated the code and it looks fine now.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 19]
```diff
@@ -1,6 +1,59 @@
...
+  text-align: center;
+  padding-top: em(8);
+  font-size: 20px;
+  position: absolute;
+  z-index: 10;
+  top: 98;
+  right: 150;
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: top and right value has to be in em

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 20]
```diff
@@ -1,6 +1,59 @@
 @import "../../../assets/stylesheets/color-variables";
 @import "../../../assets/stylesheets/custom";
 
+.add-detailed-wi{
+  text-align: center;
+  padding-top: em(8);
+  font-size: 20px;
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: font-size has to be in em. the function em(20); needs to be here

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 21]
```diff
@@ -1,6 +1,59 @@
...
+  position: absolute;
+  z-index: 10;
+  top: 98;
+  right: 150;
+  width: em(26);
+	height: em(26);
+	background: #0088ce;
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: indentation missing.

File: src/app/work-item/work-item-list/work-item-list.component.html [hunk 22]
```diff
@@ -1,5 +1,24 @@
-<div class="detail-list-container" [class.containerPadd-nologin] = "!loggedIn" >
-  
+<div *ngIf="loggedIn && showTypesModal" class="types-modal">  
+  <h5>Click to select a type</h5>
+  <ul>
+    <li *ngFor="let type of workItemTypes" (click)="onChangeType(type.name)">
+      <a>
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list.component.html]: Don't use a tag if no need of href attribute. Its behavior may change in different browser.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 23]
```diff
@@ -1,6 +1,59 @@
...
+  color: #ffffff;
+  cursor: pointer;
+	-moz-border-radius: em(13);
+	-webkit-border-radius: em(13);
+	border-radius: em(13);  
+}
+.types-modal{
+  background-color: white;
+  position: absolute;
+  top: 95;
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: top value unit missing.

File: src/app/work-item/work-item-list/work-item-list.component.html [hunk 24]
```diff
@@ -1,5 +1,22 @@
-<div class="detail-list-container" [class.containerPadd-nologin] = "!loggedIn" >
-  
+<div *ngIf="loggedIn && showTypesModal" class="types-modal">  
+  <h5>Click to select a type</h5>
+  <ul>
+    <li *ngFor="let type of workItemTypes" (click)="onChangeType(type.name)">      
+      <span almIcon [iconType]="type.name" class="xl-font"></span>
+      <br />
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.html]: Avoid <br/ >. We need to align it with width and float

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 25]
```diff
@@ -44,7 +92,7 @@
   cursor: pointer;
 }
 
-@media (max-width: ($grid-float-breakpoint - 1)){
+@media (max-width: ($grid-float-breakpoint - 1)){  
   .containerPadd-nologin{ padding-bottom: 0; }
   .workItem_Overlay{
     background-color:$color-pf-black;
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: overlay background has to translucent

**Comment by SMahil** [comment]
@nimishamukherjee @joshuawilson Don't we need common dialog/modal layout across the project rather than creating modal again and again.

I have already used ng2-modal for other dialog. Even If we want to create our own dialog we need something like that

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 26]
```diff
@@ -1,6 +1,54 @@
...
+        height: em(180);
+        border: 1px solid #ededed;
+        text-align: center;
+        margin: em(20) em(10);
+        padding-top: em(50);
+        color: #000000;
+        cursor: pointer;
+      }
+    li:hover {
+      color:  #0088ce;
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: color name has to be variable

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 27]
```diff
@@ -1,6 +1,54 @@
...
+  text-align: center;
+  padding-top: em(8);
+  font-size: em(20);
+  position: absolute;
+  z-index: 10;
+  top: em(59);
+  right: em(100);
+  width: em(26);
+	height: em(26);
+	background: #0088ce;
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: color names has to be variables

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 28]
```diff
@@ -1,6 +1,54 @@
 @import "../../../assets/stylesheets/color-variables";
 @import "../../../assets/stylesheets/custom";
 
+.add-detailed-wi{
+  text-align: center;
+  padding-top: em(8);
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: instead of padding-top we need to align it vertically and horizontally center. use this css for span

position:absolute; top:50%; left: 50%; @include(transform: translate(-50%,-50%));

File: src/app/work-item/work-item-list/work-item-list.component.html [hunk 29]
```diff
@@ -1,5 +1,22 @@
-<div class="detail-list-container" [class.containerPadd-nologin] = "!loggedIn" >
-  
+<div *ngIf="loggedIn && showTypesModal" class="types-modal">  
+  <h5>Click to select a type</h5>
+  <ul>
+    <li *ngFor="let type of workItemTypes" (click)="onChangeType(type.name)">      
+      <span almIcon [iconType]="type.name" class="xl-font"></span>
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.html]: both of these span needs to be in some div to align it in center

File: src/app/work-item/work-item-list/work-item-list.component.html [hunk 30]
```diff
@@ -1,5 +1,22 @@
-<div class="detail-list-container" [class.containerPadd-nologin] = "!loggedIn" >
-  
+<div *ngIf="loggedIn && showTypesModal" class="types-modal">
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.html]: I am not sure if this is a modal.

File: src/app/work-item/work-item-list/work-item-list.component.scss (hunk 25 above)
Comment (nimishamukherjee) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: This change is not related to this PR. We need to raise another issue to fix this.

**Comment by nimishamukherjee** [comment]
[test]

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 31]
```diff
@@ -1,6 +1,59 @@
...
+  ul{
+    list-style: none;    
+    li{      
+        float: left;
+        width: em(150);
+        height: em(180);
+        border: 1px solid $color-pf-black-200;
+        text-align: center;
+        margin: em(20) em(10);
+        padding-top: em(50);
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: the content inside li should be aligned with transform rather than padding-top.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 32]
```diff
@@ -1,6 +1,59 @@
...
+	-webkit-border-radius: em(13);
+	border-radius: em(13); 
+  span{
+    position:absolute; 
+    top:50%; 
+    left: 50%; 
+    @include transform (translate(-50%,-50%));
+  } 
+}
+.types-modal{
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: is this a modal?? or a normal layover

Comment (nimishamukherjee) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: It is a normal layover.

File: src/app/work-item/work-item-list/work-item-list.component.scss (hunk 31 above)
Comment (nimishamukherjee) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: When I try variations of transform: translate(-50%, -50%);  the li alignment goes wrong. I think the outer div and ul need to be set first. We can create a bug/enhancement to track this.

File: src/app/work-item/work-item-list/work-item-list.component.scss (hunk 18 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: Line numbers 12, 13, 16, 17, 18 are not looking indented.
